File: src/simulation/space/resource/dependency/CoefficientDependency.java
package simulation.space.resource.dependency;

import simulation.space.tile.Tile;
import simulation.space.resource.Resource;

public abstract class CoefficientDependency implements ResourceDependency {
    private double deprivationCoefficient;

    public CoefficientDependency(double deprivationCoefficient) {
        this.deprivationCoefficient = deprivationCoefficient;
    }

    @Override
    public double satisfactionPercent(Tile tile, Resource resource) {
        double result = satisfaction(tile, resource);
        return result + (1 - result) / deprivationCoefficient;
    }

    abstract double satisfaction(Tile tile, Resource resource);
}

File: src/simulation/space/resource/dependency/CoefficientDependency.kt
package simulation.space.resource.dependency

import simulation.space.resource.Resource
import simulation.space.tile.Tile


abstract class CoefficientDependency(private val deprivationCoefficient: Double) : ResourceDependency {
    override fun satisfactionPercent(tile: Tile, resource: Resource): Double {
        val result = satisfaction(tile, resource)

        return result + (1 - result) / deprivationCoefficient
    }

    abstract fun satisfaction(tile: Tile, resource: Resource): Double
}
